From e0459b1e97a98800877ee919472c907e1fb26748 Mon Sep 17 00:00:00 2001
From: Harald Hoyer <harald@redhat.com>
Date: Fri, 26 Jul 2013 09:32:52 +0200
Subject: [PATCH] fips: cope with module aliases, when checking modules

Also do not fail, if module aliases try to load CPU specific modules
like crc32c_intel.
---
 modules.d/01fips/fips.sh | 24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/modules.d/01fips/fips.sh b/modules.d/01fips/fips.sh
index d061dfe..8d8fc73 100755
--- a/modules.d/01fips/fips.sh
+++ b/modules.d/01fips/fips.sh
@@ -69,6 +69,10 @@ do_rhevh_check()
 
 do_fips()
 {
+    local _v
+    local _s
+    local _v
+    local _module
     info "Checking integrity of kernel"
     KERNEL=$(uname -r)
 
@@ -88,11 +92,25 @@ do_fips()
     FIPSMODULES=$(cat /etc/fipsmodules)
 
     info "Loading and integrity checking all crypto modules"
-    for module in $FIPSMODULES; do
-        if [ "$module" != "tcrypt" ]; then
-            modprobe ${module} || return 1
+    mv /etc/modprobe.d/fips.conf /etc/modprobe.d/fips.conf.bak
+    for _module in $FIPSMODULES; do
+        if [ "$_module" != "tcrypt" ]; then
+            if ! modprobe "${_module}"; then
+                # check if kernel provides generic algo
+                _found=0
+                while read _k _s _v; do
+                    [ "$_k" != "name" -a "$_k" != "driver" ] && continue
+                    [ "$_k" = "driver" ] && _v=$(str_replace "$_v" "_" "-")
+                    [ "$_v" != "$_module" ] && continue
+                    _found=1
+                    break
+                done </proc/crypto
+                [ "$_found" = "0" ] && return 1
+            fi
         fi
     done
+    mv /etc/modprobe.d/fips.conf.bak /etc/modprobe.d/fips.conf
+
     info "Self testing crypto algorithms"
     modprobe tcrypt || return 1
     rmmod tcrypt
